app/controllers/UploadSupportingMaterialMultipleController.scala

```diff
@@ -62,55 +63,78 @@ class UploadSupportingMaterialMultipleController @Inject()(
+
+    dataCacheConnector
+      .save(userAnswers.cacheMap)
+      .map(_ => Redirect(routes.UploadSupportingMaterialMultipleController.onPageLoad(mode)))
   }
 
-  def onSubmit(mode: Mode): Action[MultipartFormData[TemporaryFile]] = (identify andThen getData andThen requireData)
-    .async(parse.multipartFormData) { implicit request =>
+  def onFileSelected(): Action[FileAttachment] =
+    (identify andThen getData andThen requireData).async[FileAttachment](parse.json[FileAttachment]) { implicit request: DataRequest[FileAttachment] =>
+      val updatedAnswers = upsertFile(request.body, request.userAnswers)
 
-      def badRequest(mode: Mode, errorKey: String, errorMessage: String): Future[Result] = {
+      dataCacheConnector
+        .save(updatedAnswers.cacheMap)
+        .map(_ => Ok)
+    }
+
+  def renderView(mode: Mode, form: Form[String])(implicit request: DataRequest[AnyContent]) = {
+    val fileId = ju.UUID.randomUUID().toString
+
+    fileService.initiate(FileStoreInitiateRequest(
+      id = Some(fileId),
+      successRedirect = Some(routes.UploadSupportingMaterialMultipleController.onFileUploadSuccess(fileId, mode).absoluteURL),
+      errorRedirect = Some(routes.UploadSupportingMaterialMultipleController.onFileUploadError(fileId, mode).absoluteURL)
+    )).transformWith {
+      case Failure(NonFatal(_)) =>
+        Future.successful(BadGateway)
+      case Success(response) =>
         val goodsName = request.userAnswers.get(ProvideGoodsNamePage).getOrElse("goods")
-        val formWithError = form.withError(errorKey, errorMessage)
-        Future.successful(BadRequest(uploadSupportingMaterialMultiple(appConfig, formWithError, goodsName, mode)))
-      }
-
-      request.body.file("file-input").filter(_.filename.nonEmpty) match {
-        case _ if hasMaxFiles(request.userAnswers) =>
-          badRequest(mode, "validation-error", messagesApi("uploadSupportingMaterialMultiple.error.numberFiles"))
-        case Some(file) =>
-          fileService.validate(file).fold(
-            errorMessage =>
-              badRequest(mode, "file-input", errorMessage),
-            validFile => {
-              uploadFile(validFile).transformWith {
-                case Failure(NonFatal(_)) =>
-                  Future.successful(Results.BadGateway)
-                case Success(updatedAnswers) =>
-                  Future.successful(Results.Redirect(navigator.nextPage(UploadSupportingMaterialMultiplePage, mode)(updatedAnswers)))
-              }
-            }
-          )
-        case _ =>
-          badRequest(mode, "file-input", messagesApi("uploadSupportingMaterialMultiple.upload.selectFile"))
-      }
+        Future.successful(Ok(uploadSupportingMaterialMultiple(appConfig, response, form, goodsName, mode)))
     }
+  }
+
+  def onPageLoad(mode: Mode): Action[AnyContent] = (identify andThen getData andThen requireData).async { implicit request =>
+    renderView(mode, form)
+  }
```

Comment:
Need to think about what happens if the user selects a file in the picker and then goes back or refreshes the page - they will get a new UUID for the file, but their old file metadata will stay and never get uploaded